src/prop/cadical.cpp

```diff
@@ -0,0 +1,165 @@
+/*********************                                                        */
+/*! \file cadical.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Mathias Preiner
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Wrapper for CaDiCaL SAT Solver.
+ **
+ ** Implementation of the CaDiCaL SAT solver for CVC4 (bitvectors).
+ **/
+
+#include "prop/cadical.h"
+
+#ifdef CVC4_USE_CADICAL
+
+#include "proof/sat_proof.h"
+
+namespace CVC4 {
+namespace prop {
+
+using CadicalLit = int;
+using CadicalVar = int;
+
+// helper functions
+namespace {
+
+SatValue toSatValue(int result)
+{
+  if (result == 10) return SAT_VALUE_TRUE;
+  if (result == 20) return SAT_VALUE_FALSE;
+  Assert(result == 0);
+  return SAT_VALUE_UNKNOWN;
+}
+
+SatValue toSatValueLit(int value)
+{
+  if (value == 1) return SAT_VALUE_TRUE;
+  Assert(value == -1);
+  return SAT_VALUE_FALSE;
+}
+
+CadicalLit toCadicalLit(SatLiteral lit)
+{
+  return lit.isNegated() ? -lit.getSatVariable() : lit.getSatVariable();
+}
+
+CadicalVar toCadicalVar(SatVariable var) { return var; }
+
+}  // namespace helper functions
+
+CadicalSolver::CadicalSolver(StatisticsRegistry* registry,
+                             const std::string& name)
+    : d_solver(new CaDiCaL::Solver()),
+      // Note: CaDiCaL variables start with index 1 rather than 0 since negated
+      //       literals are represented as the negation of the index.
+      d_nextVarIdx(1),
+      d_statistics(registry, name)
+{
+  d_true = newVar();
+  d_false = newVar();
+
+  d_solver->set("quiet", 1);  // CaDiCaL is verbose by default
+  d_solver->add(toCadicalVar(d_true));
+  d_solver->add(0);
+  d_solver->add(-toCadicalVar(d_false));
+  d_solver->add(0);
+}
+
+CadicalSolver::~CadicalSolver() {}
+
+ClauseId CadicalSolver::addClause(SatClause& clause, bool removable)
```

Comment:
Why isn't clause `const SatClause&`? (I know, this method is inherited from SatSolver but we should probably change this at some point)

Reply:
Probably for no specific reason. We can clean this up in a separate PR later.